diff --git a/hw/xwayland/xwayland.c b/hw/xwayland/xwayland.c
index b353167c3..3afaa91a2 100644
--- a/hw/xwayland/xwayland.c
+++ b/hw/xwayland/xwayland.c
@@ -112,8 +112,11 @@ ddxUseMsg(void)
     ErrorF("-wm fd                 create X client for wm on given fd\n");
     ErrorF("-listen fd             add give fd as a listen socket\n");
     ErrorF("-eglstream             use eglstream backend for nvidia GPUs\n");
+    ErrorF("-verbose [n] verbose startup messages\n");
 }
 
+static int verbosity = 0;
+
 int
 ddxProcessArgument(int argc, char *argv[], int i)
 {
@@ -1151,6 +1154,21 @@ xwl_screen_init(ScreenPtr pScreen, int argc, char **argv)
         else if (strcmp(argv[i], "-shm") == 0) {
             xwl_screen->glamor = 0;
         }
+        else if (strcmp(argv[i], "-verbose") == 0) {
+            if (++i < argc && argv[i]) {
+                char *end;
+                long val;
+
+                val = strtol(argv[i], &end, 0);
+                if (*end == '\0') {
+                    verbosity = val;
+                    LogSetParameter(XLOG_VERBOSITY, verbosity);
+                    return 2;
+                }
+            }
+            LogSetParameter(XLOG_VERBOSITY, ++verbosity);
+            return 1;
+        }
         else if (strcmp(argv[i], "-eglstream") == 0) {
 #ifdef XWL_HAS_EGLSTREAM
             use_eglstreams = TRUE;
